projects/visual_recognition_grid_cells/SDR_CNN.py

```diff
@@ -0,0 +1,325 @@
+#  Numenta Platform for Intelligent Computing (NuPIC)
+#  Copyright (C) 2020, Numenta, Inc.  Unless you have an agreement
+#  with Numenta, Inc., for a separate license for this software code, the
+#  following terms and conditions apply:
+#
+#  This program is free software: you can redistribute it and/or modify
+#  it under the terms of the GNU Affero Public License version 3 as
+#  published by the Free Software Foundation.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#  See the GNU Affero Public License for more details.
+#
+#  You should have received a copy of the GNU Affero Public License
+#  along with this program.  If not, see http://www.gnu.org/licenses.
+#
+#  http://numenta.org/licenses/
+#
+'''
+This trains a simple supervised CNN that can be used to output SDR features 
+derived from images such as MNIST or Fashion-MNIST; these can subsequently
+be used by other classifiers including GridCellNet, or a decoder to reconstruct
+the images
+Several functions are based on the k-WTA sparse_cnn.ipynb example in 
+nupic.torch/examples
+'''
+import numpy as np
+import torch
+import os
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+from torchvision import datasets, transforms
+import matplotlib.pyplot as plt
+from nupic.torch.modules import (
+    KWinners2d, rezero_weights, update_boost_strength
+)
+import torch
+from PIL import Image
+
+torch.manual_seed(18)
+np.random.seed(18)
+
+# Parameters
+TRAIN_NEW_NET = False # To generate all the SDRs needed for down-stream use in other
```

Comment:
It might be better to make this `True` by default, since it will need to be True on first run.